apps/meteor/ee/server/hooks/messages/BeforeSaveCannedResponse.ts

```diff
@@ -0,0 +1,116 @@
+import { isILivechatVisitor, isOmnichannelRoom } from '@rocket.chat/core-typings';
+import type { IMessage, IRoom, ISetting, IUser, SettingValue } from '@rocket.chat/core-typings';
+import { License } from '@rocket.chat/license';
+import { LivechatVisitors, Users } from '@rocket.chat/models';
+import get from 'lodash.get';
+import mem from 'mem';
+
+const placeholderFields = {
+	'contact.name': {
+		from: 'visitor',
+		dataKey: 'name',
+	},
+	'contact.email': {
+		from: 'visitor',
+		dataKey: 'visitorEmails[0].address',
+	},
+	'contact.phone': {
+		from: 'visitor',
+		dataKey: 'phone[0].phoneNumber',
+	},
+	'agent.name': {
+		from: 'agent',
+		dataKey: 'name',
+	},
+	'agent.email': {
+		from: 'agent',
+		dataKey: 'emails[0].address',
+	},
+};
+
+export class BeforeSaveCannedResponse {
+	static enabled = false;
+
+	private getSetting: <T extends SettingValue = SettingValue>(id: ISetting['_id']) => T;
+
+	constructor({ getSetting }: { getSetting: <T extends SettingValue = SettingValue>(id: ISetting['_id']) => T }) {
+		this.getSetting = getSetting;
+
+		if (this.getSetting('Canned_Responses_Enable')) {
+			BeforeSaveCannedResponse.enabled = true;
+		}
+		void License.onToggledFeature('canned-responses', {
+			up: () => {
+				BeforeSaveCannedResponse.enabled = true;
+			},
+			down: () => {
+				BeforeSaveCannedResponse.enabled = false;
+			},
+		});
+	}
+
+	private getUser = mem((userId: string) => Users.findOneById(userId, { projection: { name: 1, _id: 1, emails: 1 } }), {
+		maxAge: 1000 * 30,
+	});
+
+	private getVisitor = mem((visitorId: string) => LivechatVisitors.findOneEnabledById(visitorId), {
+		maxAge: 1000 * 30,
+	});
+
+	async replacePlaceholders({
+		message,
+		room,
+		user,
+	}: {
+		message: IMessage;
+		room: IRoom;
+		user: Pick<IUser, '_id' | 'username' | 'name' | 'emails' | 'language'>;
+	}): Promise<IMessage> {
+		// If the feature is disabled, return the message as is
+		if (!BeforeSaveCannedResponse.enabled) {
```

Comment:
If it's disabled by setting, we won't be able to run it as the enabled prop will never update 😬

Reply:
I had a hard time trying to find the best solution for this..

I assumed on startup I could check the setting, which would return `false` if no EE license is applied.. then added a listener to license module to be able to enable the feature..

but I still check the setting every time on line 74, which may not be needed btw